pkg/gadgets/trace/dns/tracer/bpf/dns.c

```diff
@@ -108,23 +106,87 @@ SEC("socket1")
 int ig_trace_dns(struct __sk_buff *skb)
 {
 	struct event_t event;
+	int i, l4_off, dns_off;
 	__u16 sport, dport;
+	__u16 h_proto;
+	__u8 proto;
 	__u16 id;
-	int i;
 
-	// Skip non-IP packets
-	if (load_half(skb, offsetof(struct ethhdr, h_proto)) != ETH_P_IP)
-		return 0;
+	// Do a first pass only to extract the port and drop the packet if it's not DNS
+	h_proto = load_half(skb, offsetof(struct ethhdr, h_proto));
+	switch (h_proto) {
+	case ETH_P_IP:
+		proto = load_byte(skb,
+				  ETH_HLEN + offsetof(struct iphdr, protocol));
+		// An IPv4 header doesn't have a fixed size. The IHL field of a packet
+		// represents the size of the IP header in 32-bit words, so we need to
+		// multiply this value by 4 to get the header size in bytes.
+		__u8 ihl_byte = load_byte(skb, ETH_HLEN);
```

Comment:
How does this line deal with endianness and the [4 bits of version](https://github.com/torvalds/linux/blob/cf1182944c7cc9f1c21a8a44e0d29abe12527412/include/uapi/linux/ip.h#L88-L96)?

Why not use [the mechanism](https://github.com/inspektor-gadget/inspektor-gadget/blob/13cb07051c4617853983f2b1d3dc9c8296b8b626/pkg/gadgets/trace/network/tracer/bpf/network.c#L40-L49) the trace network gadget uses?

Reply:
This will be part of another PR.